plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/listener/SearchQueryLatencyListener.java

```diff
@@ -0,0 +1,127 @@
+
+    /**
+     * Constructor for SearchQueryLatencyListener
+     *
+     * @param clusterService The Node's cluster service.
+     * @param topQueriesByLatencyService The topQueriesByLatencyService associated with this listener
+     */
+    @Inject
+    public SearchQueryLatencyListener(ClusterService clusterService, TopQueriesByLatencyService topQueriesByLatencyService) {
+        this.topQueriesByLatencyService = topQueriesByLatencyService;
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(TOP_N_LATENCY_QUERIES_ENABLED, this::setEnabled);
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(
+                TOP_N_LATENCY_QUERIES_SIZE,
+                this.topQueriesByLatencyService::setTopNSize,
+                this.topQueriesByLatencyService::validateTopNSize
+            );
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(
+                TOP_N_LATENCY_QUERIES_WINDOW_SIZE,
+                this.topQueriesByLatencyService::setWindowSize,
+                this.topQueriesByLatencyService::validateWindowSize
+            );
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(TOP_N_LATENCY_QUERIES_EXPORTER_TYPE, this.topQueriesByLatencyService::setExporterType);
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(
+                TOP_N_LATENCY_QUERIES_EXPORTER_INTERVAL,
+                this.topQueriesByLatencyService::setExportInterval,
+                this.topQueriesByLatencyService::validateExportInterval
+            );
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(TOP_N_LATENCY_QUERIES_EXPORTER_IDENTIFIER, this.topQueriesByLatencyService::setExporterIdentifier);
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(TOP_N_LATENCY_QUERIES_EXPORTER_ENABLED, this.topQueriesByLatencyService::setExporterEnabled);
+
+        this.setEnabled(clusterService.getClusterSettings().get(TOP_N_LATENCY_QUERIES_ENABLED));
+        this.topQueriesByLatencyService.setTopNSize(clusterService.getClusterSettings().get(TOP_N_LATENCY_QUERIES_SIZE));
+        this.topQueriesByLatencyService.setWindowSize(clusterService.getClusterSettings().get(TOP_N_LATENCY_QUERIES_WINDOW_SIZE));
+    }
+
+    @Override
+    public void setEnabled(boolean enabled) {
+        super.setEnabled(enabled);
+        this.topQueriesByLatencyService.setEnableCollect(enabled);
+    }
+
+    @Override
+    public boolean isEnabled() {
+        return super.isEnabled();
+    }
+
+    @Override
+    public void onPhaseStart(SearchPhaseContext context) {}
+
+    @Override
+    public void onPhaseEnd(SearchPhaseContext context, SearchRequestContext searchRequestContext) {}
+
+    @Override
+    public void onPhaseFailure(SearchPhaseContext context) {}
+
+    @Override
+    public void onRequestStart(SearchRequestContext searchRequestContext) {}
```

Comment:
aren't we keeping request start time with topN search query record?

Reply:
The request start time is available from the context itself.